docs(filters): add initial reference (#592)
arcjet/arcjet-docs: src/content/docs/filters/reference.mdx

so that you can block certain things within Arcjet itself.

## Plan availability

Arcjet filter rules is available on any
[pricing plan](https://arcjet.com/pricing).

## Configuration

Filters are configured by allowing or denying a one or more expressions.
The `allow` and `deny` lists are mutually-exclusive,
such that using `allow` will result in a `DENY` decision for any request that
is not matched by a filter in the `allow` list and using `deny` will result in
an `ALLOW` decision for any request that is not matched by a filter in the
`deny` list.

The `arcjet` client can be configured with one or more filter rules,
which are constructed with the `filters(options)` function and configured by
`FilterOptionsAllow` or `FilterOptionsDeny`:

```ts
interface FilterOptionsAllow {
  allow: ReadonlyArray<Filter | string> | Filter | string;
  mode?: "DRY_RUN" | "LIVE" | null | undefined;
}

interface FilterOptionsDeny {
  allow: ReadonlyArray<Filter | string> | Filter | string;
  mode?: "DRY_RUN" | "LIVE" | null | undefined;
}

interface Filter {
  displayName?: string | null | undefined;
  expression: string;
}

> ```

Reviewer: Oh - once their is a release with filters there is also functionality to use `expressive-code` to reference the types directly. (I'll find it after the meeting & edit in here)

    children:
      - text: "Example: Allow non-VPN, US-based GET requests"
        anchor: "example-allow-non-vpn-us-based-get-requests"
  - text: "Expression language"
    anchor: "expression-language"
    children:
      - text: "Fields"
        anchor: "fields"
      - text: "Operators"
        anchor: "operators"
      - text: "Functions"
        anchor: "functions"
      - text: "Values"
        anchor: "values"
  - text: "Decision"
    anchor: "decision"
  - text: "Error handling"
    anchor: "error-handling"
  - text: "Testing"
    anchor: "testing"
---

import DisplayType from "@/components/DisplayType.astro";
import SlotByFramework from "@/components/SlotByFramework";
import BunAllow from "@/snippets/filters/reference/bun/Allow.mdx";
import DenoAllow from "@/snippets/filters/reference/deno/Allow.mdx";
import NestJsAllow from "@/snippets/filters/reference/nestjs/Allow.mdx";
import NextJsAllow from "@/snippets/filters/reference/nextjs/Allow.mdx";
import NodeJsAllow from "@/snippets/filters/reference/nodejs/Allow.mdx";
import RemixAllow from "@/snippets/filters/reference/remix/Allow.mdx";
import SvelteKitAllow from "@/snippets/filters/reference/sveltekit/Allow.mdx";

> Arcjet filters let you permit or block requests using

Author: Maybe just `block` instead of `permit or block`?

| `ip.src.asnum.name`      | `string`      | `"Google, LLC."`                              |
| `ip.src.asnum.type`      | `string`      | `"business"`                                  |
| `ip.src.asnum`           | `string`      | `"15169"`                                     |
| `ip.src.city`            | `string`      | `"Mountain View"`                             |
| `ip.src.continent.name`  | `string`      | `"North America"`                             |
| `ip.src.continent`       | `string`      | `"NA"`                                        |
| `ip.src.country.name`    | `string`      | `"United States"`                             |
| `ip.src.country`         | `string`      | `"US"`                                        |
| `ip.src.crawler.name`    | `string`      | `""`                                          |
| `ip.src.crawler`         | `boolean`     | `false`                                       |
| `ip.src.hosting`         | `boolean`     | `false`                                       |
| `ip.src.lat`             | `string`      | `"39.90008"`                                  |

> | `ip.src.lon`             | `string`      | `"-79.71643"`                                 |

Author: is the `lat`, `lon` still from the old fastly/Uniontown example? Does `8.8.8.8` result in different values?

| `ip.src.lon`             | `string`      | `"-121.89496"`                                |
| `ip.src.mobile`          | `boolean`     | `false`                                       |
| `ip.src.postal_code`     | `string`      | `"94043"`                                     |
| `ip.src.proxy`           | `boolean`     | `false`                                       |
| `ip.src.region`          | `string`      | `"California"`                                |
| `ip.src.relay`           | `boolean`     | `false`                                       |
| `ip.src.service`         | `string`      | `""`                                          |
| `ip.src.timezone.name`   | `string`      | `"America/Los_Angeles"`                       |
| `ip.src.tor`             | `boolean`     | `false`                                       |
| `ip.src.vpn`             | `boolean`     | `false`                                       |
| `ip.src`                 | `Ip`          | `8.8.8.8`                                     |

All `ip.src.*` fields reflect information about the client IP address (the
request source IP). These fields require data from our IP reputation database so

> a call to the Arcjet API is required. Other fields are analyzed locally.

Reviewer: ```suggestion
a call to the Arcjet Cloud API is required. Other fields are analyzed locally.
```

Reviewer 2: We'll change this in a different PR to make it consistent across the docs.

Arcjet only ever makes a single call to our Cloud API regardless of the rule

> configuration. The Arcjet API has been designed for high performance and low

Reviewer: ```suggestion
configuration. Our Cloud API has been designed for high performance and low
```

Reviewer 2: We'll change this in a different PR to make it consistent across the docs.

- `id` (`string`) - The unique ID for the request. This can be used to look up
  the request in the Arcjet dashboard. It is prefixed with `req_` for decisions
  involving the Arcjet cloud API. For decisions taken locally, the prefix is
  `lreq_`.
- `conclusion` (`ArcjetConclusion`) - The final conclusion based on evaluating
  each of the configured rules. If you wish to accept Arcjet's recommended
  action based on the configured rules then you can use this property.
- `reason` (`ArcjetReason`) - An object containing more detailed
  information about the conclusion.
- `results` (`ArcjetRuleResult[]`) - An array of `ArcjetRuleResult` objects
  containing the results of each rule that was executed.
- `ip` (`ArcjetIpDetails`) - An object containing Arcjet's analysis of the
  client IP address. See the SDK reference for more information.

You check if a deny conclusion has been returned by a filter rule by using
`decision.isDenied()` and `decision.reason.isFilterRule()` respectively.

You can iterate through the results and check whether a filter rule was applied:

```ts
for (const result of decision.results) {
  console.log("Rule Result", result);
}
```

> 

Reviewer: This, while very useful, feels slightly out of place to me. Is this maybe something that could be a callout to go see a central page about handling decisions? I feel as though most people who get to filters will be familiar

Author: I think I agree: I copied it over from bots. Looks like several such pages have this? My preferred place to end up in would be dedicated pages for such things (decision, error handling, testing) instead of copy/pasting them over?

@davidmytton what are your thoughts?

Reviewer 2: I'd like each reference to stand by itself because we see traffic landing on these guides where they probably won't have context across the product. We could consolidate into a snippet that can be included, or reference another page though. For another PR I think.

## Error handling

Arcjet is designed to fail open so that a service issue or misconfiguration does
not block all requests. The SDK will also time out and fail open after 500ms
when `NODE_ENV` is `production` and 1000ms otherwise. However, in most cases,
the response time will be less than 20-30ms.

If there is an error condition when processing the rule, Arcjet will return an
`ERROR` result for that rule and you can check the `message` property on the rule's
error result for more information.

> 

Reviewer: This could perhaps be more targeted at finding errors in filters specifically?

Reviewer: I'll resolve for a similar reason as above
